network/src/dms.rs

```diff
@@ -1,65 +1,151 @@
+use super::Storage;
 use super::*;
-use async_trait::async_trait;
-use simperby_common::{crypto::*, BlockHeight};
+use anyhow::anyhow;
+use serde_tc::http::*;
+use serde_tc::{serde_tc_full, StubCall};
+use simperby_common::*;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+#[derive(Debug, Clone, Serialize)]
+pub struct Message {
+    data: String,
+    signature: TypedSignature<String>,
+}
+
+impl Message {
+    pub fn new(data: String, signature: TypedSignature<String>) -> Result<Self, CryptoError> {
+        signature.verify(&data)?;
+        Ok(Self { data, signature })
+    }
+
+    pub fn data(&self) -> &str {
+        &self.data
+    }
+
+    pub fn signature(&self) -> &TypedSignature<String> {
+        &self.signature
+    }
+}
+
+impl ToHash256 for Message {
+    fn to_hash256(&self) -> Hash256 {
+        Hash256::hash(serde_json::to_vec(self).unwrap())
+    }
+}
+
+/// A message before verification.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct RawMessage {
+    pub data: String,
+    pub signature: TypedSignature<String>,
+}
+
+impl RawMessage {
+    #[allow(dead_code)]
+    pub fn into_message(self) -> anyhow::Result<Message> {
+        Message::new(self.data, self.signature).map_err(|e| anyhow!(e))
+    }
+
+    #[allow(dead_code)]
+    pub fn from_message(message: Message) -> Self {
+        RawMessage {
+            data: message.data().to_owned(),
+            signature: message.signature().clone(),
+        }
+    }
+}
 
 /// The interface that will be wrapped into an HTTP RPC server for the peers.
-#[async_trait]
-pub trait DistributedMessageSetRpcInterface {
+#[serde_tc_full]
+trait DistributedMessageSetRpcInterface: Send + Sync + 'static {
     /// Returns the messages except `knowns`. If the height is different, it returns `Err(height)`.
```

Comment:
1. It seems `DmsRpcInterface::get_message` (at line 63) doesn't return `Err(height)` anymore.
2. Why did you change the error type?

Reply:
1. Fixed.
2. I found that there could be more error cases besides height mismatch.

Reply:
alright